### PR TITLE
💫 Add TrueUSD and TokenController for BSC Chain

### PR DESCRIPTION
**Pull request by qiwihui**

### CODE REVIEW

**Review by msieczko — CHANGES_REQUESTED** [review]
I suggest moving the files from `contracts/bsc` to `packages/contracts-bsc` for consistency with the latest practice in this repo. I would copy the dependencies and lint/build scripts setup from `packages/contracts-por` as well.

I also think that it would be good to start documenting the versions of code that are deployed in a way did it in PR #1235. It will make future reviews of deployed code easier.

CC: @xiaomizhou @tt-krzysztof

**Comment by yuchenlintt** [comment]
Hi @qiwihui! I hope you're doing well. Thank you for your work in preparing this PR. And my apologies in advance for leaving you this long review comment :)

My team has spent ~1 day reviewing your proposed changes. So far we have not found any major issues, and we believe most of your diffs (which we have attempted to manually check against corresponding files in `packages/contracts-por`) look fairly sensible.

Nonetheless, given the magnitude of money ($1B+) that these contracts will potentially be responsible for, as well as the high consequences of making even a single-line mistake, we are quite concerned about:
1. the lack of clear commit history connecting `packages/contracts-bsc` back to `packages/contracts-por`,
2. the lack of code review and deliberate justification for individual changes, and
3. the lack of unit tests to verify that there have been no regressions in functionality and/or safety.

I will elaborate on each point below and then point you to some next steps that we would like to see.

### Commit history

The original TUSD code has undergone several rounds of audit by several parties. It has also been battle-hardened by handling $1B+ in assets on Ethereum mainnet, in full view of bug bounties, white hats, and black hats, for several years.

However, whatever confidence we gain from this scrutiny only applies to this original code in `main` (currently 979aeabacbc9a81916356f8d2e8b158ba6a33224), along a commit history that we can trace back to all previously scrutinized versions.

Your first commit in this PR (1b0fe299fe17265680baeec40e7180ce6803881d) contains files that are copied + unmodified (ideal!), files that are copied + renamed (okay, but not optimal), files that are copied + modified internally (dangerous), and files that are inlined + deleted (also dangerous).

Since your PR's first commit does not extend this commit history, we needed to manually check for diffs in every single file between the original code in `main` and your first commit in this PR. The large gap between these two commits raises the likelihood that, during our manual check, we overlooked a small change that accidentally introduced a critical vulnerability.

### Individual changes

In our typical development practice, we expect that each PR contains one isolated change that is easy to review independently of all other irrelevant changes. We might relax this requirement when modifying unit tests, but changes to live Solidity code are especially dangerous, so we tend to enforce this fairly strictly.

Here is a good example of one such PR: https://github.com/trusttoken/contracts-pre22/pull/1221. Notice how this PR documents how and why we made this change (we're deprecating an unused modifier and constant; this unused code had been found during an audit and assigned ID `A1`), and it does not introduce irrelevant refactors.

Even if a refactor would make the code cleaner or simpler overall, we generally do not allow it _unless_ it can be separately reviewed in its own PR. This helps us make sure we're reviewing each change carefully on its own merits.

Overall, we acknowledge that this approach means that the TrueCurrency contracts update slowly, still use outdated design patterns, and have numerous possibilities for improvement -- the vast majority of these non-optimal design patterns had already been stuck this way before I joined Archblock! -- but the tradeoff is that at least we know we aren't introducing vulnerabilities to code that has already been very carefully examined.

### Unit tests

There is an existing suite of unit tests in `packages/contracts-por/test/`. It is already less testing than I would prefer for smart contracts with this much at stake; however, we don't at present have the engineering capacity to add more tests, so at minimum we will need to make do with the tests we already have.

But I noticed that you had not copied over these unit tests to `packages/contracts-bsc`. By itself that wouldn't necessarily be an issue. If at any time we could easily perform a diff of `packages/contracts-por/contracts/` against `packages/contracts-bsc/contracts/` and confirm that there are zero lines of code changed, then we can be reasonably certain that the contracts in `contracts-bsc` would still pass the unit tests in `contracts-por`.

However, it is the combination of {making changes to code} _and_ {not keeping unit tests} that makes me concerned that a regression could have been introduced.

## Next steps

I propose we recreate this PR's changes in a parallel queue of individually-reviewed PRs. If you would like, you are welcome to set the base of each PR to be the branch below it, so that these form a queue that can be submitted for review all at once and then later merged one at a time. (Note that if you do so, you may have to resolve several merge conflicts if we request changes to a PR in the middle of your queue.)

1. I recommend you keep this branch `tutoken:tron-BSC-pr` and PR #1233 open for now. It may be useful for you to look at this branch and PR as a reference.
2. For the first parallel PR, I recommend you copy the directory `packages/contracts-por` over to `packages/contracts-bsc` verbatim. You may update the `package.json` info, and, if necessary, fix the build and any failing unit tests. This ensures that the commit history logs the exact lineage of the TUSD code.
3. For subsequent PRs, please make individual changes as noted above. We are able to review stylistic/design changes, but it may be in your best interest to stay focused on the business logic so this process doesn't take too long.
4. Once we've merged the whole parallel queue of PRs and confirmed that it deploys to BSC, you may delete this PR #1233.

I hope this helps clarify our approach to Solidity code review! And thank you again for your PR submission.

**Comment by qiwihui** [comment]
Hi @yuchenlintt, thank you very much for your review against the code. As you have pointed out, the first commit is quite messy since it is is squashed from several commits from the original code base. It is quite inconvenient to review and test.
So I have submitted another pull request https://github.com/trusttoken/contracts-pre22/pull/1237 which contains the detailed commit history from `packages/contracts-por` to `packages/contracts-bsc` and fixes unit tests for the contracts. And once this PR has been merged, I will some other PRs based on your comments.

File: packages/contracts-bsc/contracts/BscTokenController.sol [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+    }
+
+    /*
+    ========================================
+    proxy functions
+    ========================================
+    */
+
+    function transferTrueCurrencyProxyOwnership(address _newOwner) external onlyOwner {
+        OwnedUpgradeabilityProxy(address(uint160(address(token)))).transferProxyOwnership(_newOwner);
```
[the remark this answers is not in the document]
Comment (qiwihui) [reply packages/contracts-bsc/contracts/BscTokenController.sol:254]: When using `OUP(address(token))`, a TypeError will be raised with the message 'Explicit type conversion not allowed from non-payable "address" to "contract OwnedUpgradeabilityProxy", which has a payable fallback function.' Therefore, to avoid this error, you can either use `OwnedUpgradeabilityProxy(address(uint160(address(token))))` or use the interface `IOwnedUpgradeabilityProxy(address(token))`.

**Comment by qiwihui** [comment]
Hello @yuchenlintt and @msieczko, I have reviewed the contracts on the main branch. The necessary changes have been merged. Additionally, all test cases for BSC have passed successfully. Therefore, I believe we can mark this issue as completed.

I would like to express my gratitude to @yuchenlintt once again for helping me review the contract and providing valuable suggestions.
